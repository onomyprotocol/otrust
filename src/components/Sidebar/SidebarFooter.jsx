import React from 'react';
import styled from 'styled-components';

import { responsive } from 'theme/constants';
import { MediumIcon, TwitterIcon } from './SidebarIcons';

const SidebarFooterWrapper = styled.footer`
  display: flex;
  align-items: center;
  gap: 16px;

  padding: 50px 40px;
  margin-top: auto;

  @media screen and (max-width: ${responsive.laptop}) {
    padding: 40px 24px;
  }

  @media screen and (max-width: ${responsive.tablet}) {
    width: 100%;
    gap: 12px;

    position: absolute;
    bottom: 0;
    left: 0;
  }

  @media screen and (max-width: ${responsive.smartphoneLarge}) {
    padding: 24px 20px;

    background-color: ${props => props.theme.colors.bgDarken};
  }
`;

const Link = styled.a`
  display: block;
  margin-right: auto;

  color: ${props => props.theme.colors.textSecondary};
  text-decoration: none;

  &:hover {
    color: ${props => props.theme.colors.textSecondary_lighten};
  }

  &:active {
    color: ${props => props.theme.colors.textSecondary_darken};
  }
`;

const SecondaryIcon = styled.a`
  display: flex;
  align-items: center;
  justify-content: center;

  height: 40px;
  width: 40px;

  background-color: ${props => props.theme.colors.bgHighlightBorder};
  border-radius: 8px;

  font-size: 20px;
  color: ${props => props.theme.colors.iconsSecondary};

  cursor: pointer;

  svg {
    width: 20px;
    height: 20px;
  }

  @media screen and (max-width: ${responsive.laptop}) {
    width: 32px;
    height: 32px;

    font-size: 14px;

    svg {
      width: 16px;
      height: 16px;
    }
  }

  &:hover {
    background-color: ${props => props.theme.colors.bgHighlightBorder_lighten};
  }

  &:active {
    background-color: ${props => props.theme.colors.bgHighlightBorder_darken};
  }
`;

export default function SidebarFooter() {
  return (
    <SidebarFooterWrapper>
      <Link href="https://onomy.io/" target="_blank">
        About Onomy
      </Link>

      <SecondaryIcon href="https://medium.com/onomy-protocol" target="_blank">
        <MediumIcon />
      </SecondaryIcon>

      <SecondaryIcon href="https://twitter.com/onomyprotocol" target="_blank">
        <TwitterIcon />
      </SecondaryIcon>
    </SidebarFooterWrapper>
  );
}
