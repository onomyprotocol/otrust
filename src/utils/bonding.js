const a = 100000000;

export function priceAtSupply(supply) {
  return (supply / a) ** 2;
}

// Bonding curve price at specified supply
export function supplyAtPrice(price) {
  //#NOM Sold = sqrt(ETH/NOM) * a
  return Math.sqrt(price) * a;
}

// NOM supply range to ETH
export function NOMsupplyETH(supplyTop, supplyBot) {
  // Integrate over curve to get amount of ETH needed to buy amount of NOM
  // ETH = a/3((supplyNOM_Top/a)^3 - (supplyNOM_Bot/a)^3)
  return (a / 3) * ((supplyTop / a) ** 3 - (supplyBot / a) ** 3);
}

// Returns Buy Quote for the purchase of NOM based on amount of ETH
// Parameters:
// Input
// amount: amount of ETH
// Output
// output: amount of NOM
export function ETHtoNOM(amount, supply) {
  const price = priceAtSupply(supply);
  // 1. Determine supply bottom
  // Price bottom is 1% above priceBondCurve
  const priceBot = price * 1.01;
  const supplyBot = supplyAtPrice(priceBot);
  // 2. Integrate over curve, and solve for supply top
  // (3*ETH/a + (supplyNOM_Bot/a)^3)^(1/3)
  const supplyTop = a * ((3 * amount) / a + supplyBot ** 3 / a ** 3) ** (1 / 3);
  // 3. Subtract supply bottom from top to get amount NOM for amount ETH
  const diff = supplyTop - supplyBot;
  return { supplyBot, supplyTop, diff };
}

// Returns Sell Quote: NOM for ETH (Dec 18)
// Parameters:
// Input
// uint256 amountNOM: amount of NOM to be sold (18 decimal)
// Output
// uint256: amount of ETH paid in Wei or ETH (18 decimal)
export function NOMtoETH(amount, supplyTop) {
  const supplyBot = supplyTop - amount;
  const diff = 0.99 * NOMsupplyETH(supplyTop, supplyBot);
  return { supplyBot, supplyTop, diff };
}
