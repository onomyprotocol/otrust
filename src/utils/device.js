const size = {
  mobileS: '320px',
  mobileM: '375px',
  mobileL: '425px',
  tablet: '768px',
  laptop: '1024px',
  desktopS: '1366px',
  desktopM: '1920px',
  desktopL: '2560px',
};

export const device = {
  mobileS: `(min-width: ${size.mobileS})`,
  mobileM: `(min-width: ${size.mobileM})`,
  mobileL: `(min-width: ${size.mobileL})`,
  tablet: `(min-width: ${size.tablet})`,
  laptop: `(min-width: ${size.laptop})`,
  desktopS: `(min-width: ${size.desktopS})`,
  desktopM: `(min-width: ${size.desktopM})`,
  desktopL: `(min-width: ${size.desktopL})`,
};
