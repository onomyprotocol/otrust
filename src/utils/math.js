import { BigNumber } from 'bignumber.js';

export function chopFloat(value, precision) {
  return Math.round(value * 10 ** precision) / 10 ** precision;
}

export function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

export function truncate(str, maxDecimalDigits) {
  if (str.includes('.')) {
    const parts = str.split('.');
    return parts[0] + '.' + parts[1].slice(0, maxDecimalDigits);
  }
  return str;
}

export function format18(bignumber) {
  return bignumber.div(new BigNumber(10 ** 18));
}

export function parse18(bignumber) {
  return bignumber.times(new BigNumber(10 ** 18));
}

export var isNumber = function isNumber(value) {
  return typeof value === 'number' && isFinite(value);
};
