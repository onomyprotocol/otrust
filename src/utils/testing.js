import { ThemeProvider } from 'styled-components';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';

import { darkNew } from 'Theme/theme';
import { ChainContext } from '../context/chain/ChainContext';
import { ExchangeContext } from '../context/exchange/ExchangeContext';

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
  cache: new InMemoryCache(),
});

export const renderWithTheme = (Component, props, children) => {
  if (children) {
    return render(
      <ThemeProvider theme={darkNew}>
        <Component {...props}>{children}</Component>
      </ThemeProvider>,
    );
  }
  return render(
    <ThemeProvider theme={darkNew}>
      <Component {...props} />
    </ThemeProvider>,
  );
};

export const renderWithContext = (Component, props) => {
  return render(
    <ThemeProvider theme={darkNew}>
      <ApolloProvider client={client}>
        <ChainContext.Provider value={{ supplyNOM: new BigNumber(0), theme: darkNew }}>
          <ExchangeContext.Provider value={{ askAmount: 0, bidAmount: 0, bidDenom: 'strong' }}>
            <Component {...props} />
          </ExchangeContext.Provider>
        </ChainContext.Provider>
      </ApolloProvider>
    </ThemeProvider>,
  );
};
