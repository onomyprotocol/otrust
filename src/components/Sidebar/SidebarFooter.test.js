import { cleanup } from '@testing-library/react';

import SidebarFooter from './SidebarFooter';
import { renderWithTheme } from '../../utils/testing';

describe('Given the SidebarFooter component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(SidebarFooter);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
