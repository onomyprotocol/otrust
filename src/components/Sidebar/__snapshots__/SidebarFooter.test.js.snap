// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the SidebarFooter component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <footer
    class="sc-bdnxRM khlXFm"
  >
    <a
      class="sc-gtsrHT hlaaip"
      href="https://onomy.io/"
      target="_blank"
    >
      About Onomy
    </a>
    <a
      class="sc-dlnjwi ybHLZ"
      href="https://medium.com/onomy-protocol"
      target="_blank"
    >
      <svg
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          opacity="0.5"
        >
          <path
            d="M0 0h20v20H0z"
            fill="none"
          />
          <path
            d="M11.281 9.979A5.641 5.641 0 115.641 4.3a5.66 5.66 0 015.641 5.679m6.188 0c0 2.952-1.263 5.347-2.82 5.347s-2.82-2.394-2.82-5.347 1.261-5.346 2.818-5.346 2.82 2.394 2.82 5.347m2.531 0c0 2.645-.444 4.79-.992 4.79s-.992-2.145-.992-4.79.445-4.79.993-4.79S20 7.334 20 9.979"
            fill="#8d89a3"
          />
        </g>
      </svg>
    </a>
    <a
      class="sc-dlnjwi ybHLZ"
      href="https://twitter.com/onomyprotocol"
      target="_blank"
    >
      <svg
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h20v20H0z"
          fill="none"
        />
        <path
          d="M16.555 6.438v.46a10.4 10.4 0 01-.66 3.6 10.744 10.744 0 01-1.961 3.282 9.729 9.729 0 01-3.2 2.4 10.445 10.445 0 01-4.383.921 10.247 10.247 0 01-2.922-.42 9.937 9.937 0 01-2.6-1.181q.42.041.86.04a7.14 7.14 0 002.381-.4 7.457 7.457 0 002.081-1.121 3.592 3.592 0 01-2.081-.73 3.463 3.463 0 01-1.261-1.771 3.579 3.579 0 00.66.06 3.7 3.7 0 00.941-.12 3.472 3.472 0 01-1.461-.67 3.567 3.567 0 01-1.031-1.261 3.749 3.749 0 01-.39-1.591v-.06a3.369 3.369 0 001.641.46 3.617 3.617 0 01-1.171-1.291 3.517 3.517 0 01-.43-1.711 3.574 3.574 0 01.48-1.8 10.262 10.262 0 007.4 3.762 4.156 4.156 0 01-.08-.82 3.5 3.5 0 01.48-1.8 3.627 3.627 0 011.311-1.311 3.462 3.462 0 011.791-.49 3.563 3.563 0 012.622 1.141 6.994 6.994 0 002.281-.881 3.481 3.481 0 01-.61 1.151 3.694 3.694 0 01-.971.851 7.316 7.316 0 002.081-.58 7.525 7.525 0 01-1.798 1.881z"
          fill="#8d89a3"
          opacity="0.5"
        />
      </svg>
    </a>
  </footer>
</DocumentFragment>
`;
